features/FEATURE_BLE/ble/gap/ConnectionParameters.h

```diff
@@ -0,0 +1,300 @@
+            _enabledPhy[ble::phy_t::LE_2M],
+            _enabledPhy[ble::phy_t::LE_CODED]
+        );
+        return set.value();
+    }
+
+    /* these return pointers to arrays of settings valid only across the number of active PHYs */
+
+    const uint16_t* getScanIntervalArray() const {
+        return &_scanInterval[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getScanWindowArray() const {
+        return &_scanWindow[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getMinConnectionIntervalArray() const {
+        return &_minConnectionInterval[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getMaxConnectionIntervalArray() const {
+        return &_maxConnectionInterval[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getSlaveLatencyArray() const {
+        return &_slaveLatency[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getConnectionSupervisionTimeoutArray() const {
+        return &_connectionSupervisionTimeout[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getMinEventLengthArray() const {
+        return &_minEventLength[getFirstEnabledIndex()];
+    }
+
+    const uint16_t* getMaxEventLengthArray() const {
+        return &_maxEventLength[getFirstEnabledIndex()];
+    }
+
+private:
+    uint8_t getFirstEnabledIndex() const {
+        if (_enabledPhy[ble::phy_t::LE_1M]) {
+            return 0;
+        } else if (_enabledPhy[ble::phy_t::LE_2M]) {
+            return 1;
+        } else if (_enabledPhy[ble::phy_t::LE_CODED]) {
+            return 2;
+        }
+        /* this should never happen, it means you were trying to start a connection with a blank set
+         * of paramters - you need to enabled at least one phy */
+        MBED_ASSERT("Trying to use connection parameters without any PHY defined.");
+        return 0;
+    }
+
+    /** Handle toggling PHYs on and off and return the correct index to use to set the configuration elements.
+     *
+     * @param phy Which Phy is being toggle.
+     * @param enable On or Off.
+     * @return The index to the array of settings.
+     */
+    uint8_t handlePhyToggle(ble::phy_t phy, bool enable) {
+        uint8_t index = phy.value();
+
+        bool was_swapped = false;
+        bool is_swapped = false;
+
+        if (isSwapped()) {
+            was_swapped = true;
+        }
+
+        _enabledPhy[phy.value()] = enable;
+
+        if (isSwapped()) {
+            is_swapped = true;
+        }
+
+        if (was_swapped != is_swapped) {
+            swapCodedAnd2M();
+        }
+
+        if (is_swapped && phy == ble::phy_t::LE_CODED) {
+            index -= 1;
+        }
+
+        return index;
+    }
+
+    bool isSwapped() const {
+        return (_enabledPhy[ble::phy_t::LE_1M] &&
+                !_enabledPhy[ble::phy_t::LE_2M] &&
+                _enabledPhy[ble::phy_t::LE_CODED]);
+    }
+
+    /** Handle the swapping of 2M and CODED so that the array is ready for the pal call. */
+    void swapCodedAnd2M() {
```

Comment:
Seems a bit convoluted? What's the rationale behind the way these arrays are handled? And why we have multiple arrays of variables instead of a single array of structs?

Reply:
because the pal call requires the parameters to be arranged in memory like that - arrays of parameters for each phy, and they need to be contiguous even if there are only 2 of them - hence the swapping

Reply:
@donatieng I'm resolving this. I think everyone agree it ain't pretty but it is an implementation detail. We can rework it latter if required; it will just cost 48 bytes of stack.